pkg/plugins/k8s/k8s_plugin.go

```diff
@@ -0,0 +1,336 @@
+	SpecVersion string
+}
+
+type McoService struct {
+	Name     string
+	Contents string
+}
+
+type McoFile struct {
+	Path     string
+	Contents struct {
+		Inline string
+	}
+}
+
+type McoServiceInjection struct {
+	Name    string
+	Dropins []struct {
+		Contents string
+	}
+}
+
+type k8sUpdateTarget struct {
+	switchdevService bool
+	switchdevScript  bool
+	networkManager   bool
+	openVSwitch      bool
+}
+
+func (u *k8sUpdateTarget) needUpdate() bool {
+	return u.switchdevService || u.switchdevScript || u.networkManager || u.openVSwitch
+}
+
+const (
+	mcoManifestPath           = "bindata/manifests/machine-config/"
+	switchdevUnits            = mcoManifestPath + "switchdev-units/"
+	switchdevUnitFile         = switchdevUnits + "switchdev-configuration.yaml"
+	networkManagerUnitFile    = switchdevUnits + "NetworkManager.service.yaml"
+	ovsUnitFile               = mcoManifestPath + "ovs-units/ovs-vswitchd.service.yaml"
+	configuresSwitchdevScript = mcoManifestPath + "files/configure-switchdev.sh.yaml"
+
+	systemdDir = "/usr/lib/systemd/system/"
+	chroot     = "/host"
+)
+
+var (
+	Plugin K8sPlugin
+
+	serviceManager service.ServiceManager
+	updateTarget   *k8sUpdateTarget
+
+	switchdevRunScript    *McoFile
+	switchdevService      *service.Service
+	openVSwitchService    *service.Service
+	networkManagerService *service.Service
+)
+
+// Initialize our plugin and set up initial values
+func init() {
+	Plugin = K8sPlugin{
+		PluginName:  "k8s_plugin",
+		SpecVersion: "1.0",
+	}
+
+	serviceManager = service.NewServiceManager(chroot)
+}
+
+// Name returns the name of the plugin
+func (p *K8sPlugin) Name() string {
+	return p.PluginName
+}
+
+// Spec returns the version of the spec expected by the plugin
+func (p *K8sPlugin) Spec() string {
+	return p.SpecVersion
+}
+
+// OnNodeStateAdd Invoked when SriovNetworkNodeState CR is created, return if need dain and/or reboot node
+func (p *K8sPlugin) OnNodeStateAdd(state *sriovnetworkv1.SriovNetworkNodeState) (needDrain bool, needReboot bool, err error) {
+	glog.Info("k8s-plugin OnNodeStateAdd()")
+	return p.OnNodeStateChange(nil, state)
+}
+
+// OnNodeStateChange Invoked when SriovNetworkNodeState CR is updated, return if need dain and/or reboot node
+func (p *K8sPlugin) OnNodeStateChange(old, new *sriovnetworkv1.SriovNetworkNodeState) (needDrain bool, needReboot bool, err error) {
+	glog.Info("k8s-plugin OnNodeStateChange()")
+	needDrain = false
+	needReboot = false
+
+	err = p.readMcoFiles()
+	if err != nil {
+		return
+	}
+
+	// Check services
+	updateTarget, err = handleServices()
+	if err != nil {
+		return
+	}
+
+	// Check switchdev config
+	var update, remove bool
+	if update, remove, err = utils.WriteSwitchdevConfFile(new); err != nil {
+		glog.Errorf("k8s-plugin OnNodeStateChange():fail to update switchdev.conf file: %v", err)
+		return
+	}
+	if remove {
+		glog.Info("k8s-plugin OnNodeStateChange(): need reboot node to clean switchdev VFs")
+		needDrain = true
+		needReboot = true
```

Comment:
I don't think we need to reboot in any of this cases.

Reply:
The update VFs in the generic plugin ignores the switchdev vfs, if there is 8 switchdev vfs we want to change them in legacy, it will be ignored because the requested num of vfs == existing num of vfs